### PR TITLE
Add printing warnings if workspace resources cannot be estimated properly

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
It's following the discussion that took place in already merged PR https://github.com/eclipse/che/pull/13185#discussion_r277561687

This PR adds printing warnings if workspace resources cannot be estimated properly.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/pull/13185#discussion_r277561687

#### Release Notes
N/A

#### Docs PR
N/A

### CODE REVIEW

File: multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/usage/tracker/RamResourceUsageTracker.java [hunk 1]
```diff
@@ -81,9 +86,18 @@ public RamResourceUsageTracker(
           if (startingEnvironment != null) {
             currentlyUsedRamMB += environmentRamCalculator.calculate(startingEnvironment);
           }
+          if (SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(config.getAttributes())) {
+            LOG.warn(
+                "Sidecar based workspace `{}` is starting, memory of its plugins is not taken in "
```
Comment (metlos) [inline multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/usage/tracker/RamResourceUsageTracker.java]: How about : ``"Memory consumption of a plugin in a sidecar-based workspace `{}` is not taken into account while the workspace is starting. The memory limits may not be applied correctly."``

File: multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/usage/tracker/RamResourceUsageTracker.java [hunk 2]
```diff
@@ -81,9 +86,18 @@ public RamResourceUsageTracker(
...
+                    + "account while workspace is starting. Set memory resources limits may work incorrectly");
+          }
+        } else {
+          // Estimation of memory for starting workspace with Devfile is not implemented yet
+          // just ignore such
+          LOG.warn(
+              "Devfile based workspace `{}` is starting, its memory is not taken in account "
```
Comment (metlos) [inline multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/usage/tracker/RamResourceUsageTracker.java]: How about: ``"Memory consumption is not taken into account while a devfile-based workspace `{}` is starting. The memory limits may not be applied correctly."``

File: multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java [hunk 3]
```diff
@@ -165,13 +169,32 @@ public WorkspaceImpl updateWorkspace(String id, Workspace update)
   }
 
   @VisibleForTesting
-  void checkMaxEnvironmentRam(WorkspaceConfig config) throws ServerException {
+  void checkMaxEnvironmentRam(Workspace workspace) throws ServerException {
+    if (workspace.getConfig() != null) {
+      checkMaxEnvironmentRam(workspace.getId(), workspace.getConfig());
+    } else {
+      LOG.warn(
+          "Can not estimate memory needed for devfile based workspace {}. Set environment memory limits may work incorrectly",
```
Comment (metlos) [inline multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java]: How about: ``"Cannot estimate the memory requirements of a devfile-based workspace `{}`. The memory limits may not be applied correctly."``

File: multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java [hunk 4]
```diff
@@ -165,13 +169,32 @@ public WorkspaceImpl updateWorkspace(String id, Workspace update)
...
+      throws ServerException {
     if (maxRamPerEnvMB < 0) {
       return;
     }
     if (config.getEnvironments().isEmpty()) {
       return;
     }
+
+    if (SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(config.getAttributes())) {
+      LOG.warn(
```
Comment (metlos) [inline multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java:194]: How about: ``"Cannot estimate the memory requirements of a devfile-based workspace{}. The memory limits may not be applied correctly."``

File: multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java [hunk 5]
```diff
@@ -193,11 +216,37 @@ void checkMaxEnvironmentRam(WorkspaceConfig config) throws ServerException {
...
   void checkRamResourcesAvailability(
-      String accountId, String namespace, WorkspaceConfig config, @Nullable String envName)
+      String accountId, String namespace, Workspace workspace, @Nullable String envName)
+      throws NotFoundException, ServerException, ConflictException {
+    if (workspace.getConfig() != null) {
+      checkRamResourcesAvailability(
+          accountId, namespace, workspace.getId(), workspace.getConfig(), envName);
+    } else {
+      LOG.warn(
+          "Can not estimate memory needed for devfile based workspace {}. Set memory resources limits may work incorrectly",
```
Comment (metlos) [inline multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java]: How about: ``"Cannot estimate the memory requirements of a devfile-based workspace `{}`. The memory limits may not be applied correctly."``

File: multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java [hunk 6]
```diff
@@ -193,11 +216,37 @@ void checkMaxEnvironmentRam(WorkspaceConfig config) throws ServerException {
...
+      WorkspaceConfig config,
+      @Nullable String envName)
       throws NotFoundException, ServerException, ConflictException {
     if (config.getEnvironments().isEmpty()) {
       return;
     }
+
+    if (SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(config.getAttributes())) {
+      LOG.warn(
+          "Sidecar memory of plugins of workspace `{}` is not taken in account. "
```
Comment (metlos) [inline multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java]: How about: ``"The memory requirements for the plugin sidecars of workspace `{}` are not taken into account. The memory limits may not be applied correctly."``

**Review by metlos — COMMENTED** [review]
LGTM barring the slight improvements to the grammar of the log messages.

**Comment by sleshchenko** [comment]
@metlos Thanks for the review. I've applied your suggestion. Can you take a look one more time?

File: multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java [hunk 7]
```diff
@@ -165,13 +169,34 @@ public WorkspaceImpl updateWorkspace(String id, Workspace update)
   }
 
   @VisibleForTesting
-  void checkMaxEnvironmentRam(WorkspaceConfig config) throws ServerException {
+  void checkMaxEnvironmentRam(Workspace workspace) throws ServerException {
+    if (workspace.getConfig() != null) {
+      checkMaxEnvironmentRam(workspace.getId(), workspace.getConfig());
+    } else {
+      LOG.warn(
+          "Can not estimate memory needed for devfile based workspace `{}`."
```
Comment (metlos) [inline multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java:177]: I forgot about this one it seems: ``"Cannot estimate the memory needed for a devfile-based workspace `{}`."``

Comment (sleshchenko) [reply multiuser/api/che-multiuser-api-resource/src/main/java/org/eclipse/che/multiuser/resource/api/workspace/LimitsCheckingWorkspaceManager.java:177]: I thought I copied/pasted your suggestion :man_facepalming: 
will fix here and recheck all others.

**Review by amisevsk — COMMENTED** [review]
Generally LGTM; one question:

How many times will the warning about not being able to figure out resources be logged? From a read of the changes, it looks like every time the "not able to estimate" situation comes up it will be logged twice or more. Could we avoid this by having one check in e.g. `createWorkspace` and avoid the others?

**Comment by sleshchenko** [comment]
It produces a lot of messages in the Server Logs, the solution is not going to be improved. So, just close it.

**Pull request closed without merging**